File: yolov8/inference_tflite.cpp
#include <iostream>
#include <algorithm>
#include <chrono>

#include <opencv2/opencv.hpp>
#include <tensorflow/lite/interpreter.h>
#include <tensorflow/lite/kernels/register.h>
#include <tensorflow/lite/model.h>
#include <tensorflow/lite/tools/gen_op_registration.h>

using namespace cv;
using namespace std;

struct Arguments
{
    string modelPath;
    string videoPath;
    int delay;
    float confidenceThreshold;
};

Arguments parseArgs(int argc, char **argv)
{
    Arguments args;
    args.modelPath = "../models/yolov8m_plates_e25.tflite";
    args.videoPath = "../datasets/plates/test_img_per_frame.mp4";
    args.delay = 1;
    args.confidenceThreshold = 0.5;

    for (int i = 1; i < argc - 1; i++) // ensure that the next argument can be always read
    {
        if (string(argv[i]) == "--model" || string(argv[i]) == "-m")
        {
            args.modelPath = argv[++i];
        }
        else if (string(argv[i]) == "--video" || string(argv[i]) == "-v")
        {
            args.videoPath = argv[++i];
        }
        else if (string(argv[i]) == "--delay" || string(argv[i]) == "-d")
        {
            args.delay = stoi(argv[++i]);
        }
        else if (string(argv[i]) == "--confidence" || string(argv[i]) == "-c")
        {
            args.confidenceThreshold = stof(argv[++i]);
        }
    }

    return args;
}

vector<cv::Rect> runInference(cv::Mat &frame, unique_ptr<tflite::Interpreter> &interpreter, float confidenceThreshold)
{
    cerr << "Running inference" << endl;
    auto start = chrono::high_resolution_clock::now();

    // image dimensions    
    int input_width = frame.cols;
    int input_height = frame.rows;

    // transform and copy the input data into the input tensor
    auto input_tensor = interpreter->inputs()[0];
    float *input = interpreter->typed_tensor<float>(interpreter->inputs()[0]);
    for (int i = 0; i < input_width * input_height; i++)
    {
        // from HWC to CHW, normalize to [0, 1]
        input[i] = frame.data[2 + i * 3] / 255.0;
        input[i + input_width * input_height] = frame.data[1 + i * 3] / 255.0;
        input[i + 2 * input_width * input_height] = frame.data[i * 3] / 255.0;
    }

    // run inference
    interpreter->Invoke();

    const float *boxes = interpreter->typed_output_tensor<float>(0);
    if (!boxes)
    {
        cerr << "Failed to retrieve output tensors." << endl;
        exit(1);
    }

    // retrieve detected boxes
    vector<cv::Rect> detected_boxes;
    int dims1 = interpreter->tensor(interpreter->outputs()[0])->dims->data[1];
    int dims2 = interpreter->tensor(interpreter->outputs()[0])->dims->data[2];
    for (int i = 0; i < dims2; i++)
    {
        if (boxes[4 * dims2 + i] > confidenceThreshold)
        {
            float s1 = boxes[i];
            float s2 = boxes[dims2 + i];
            float w = boxes[2 * dims2 + i] * 0.5;
            float h = boxes[3 * dims2 + i] * 0.5;

            // convert from center, width, height to top-left and right-bottom corners
            int x1 = s1 - w;
            int y1 = s2 - h;
            int x2 = s1 + w;
            int y2 = s2 + h;
            detected_boxes.emplace_back(cv::Rect(cv::Point(x1, y1), cv::Point(x2, y2)));
            cerr << "Detected box: " << x1 << "x" << y1 << " " << x2 << "x" << y2 << " with confidence: " << boxes[4 * dims2 + i] << endl;
        }
    }

    auto end = chrono::high_resolution_clock::now();
    cout << "Processed frame in " << chrono::duration_cast<chrono::microseconds>(end - start).count() / 1000.0f << " ms." << endl;

    return detected_boxes;
}

int main(int argc, char **argv)
{
    Arguments args = parseArgs(argc, argv);

    // load the TFLite model
    unique_ptr<tflite::FlatBufferModel> model = tflite::FlatBufferModel::BuildFromFile(args.modelPath.c_str());
    if (!model)
    {
        cerr << "Failed to load model" << endl;
        return -1;
    }

    // build the interpreter
    tflite::ops::builtin::BuiltinOpResolver resolver;
    unique_ptr<tflite::Interpreter> interpreter;
    tflite::InterpreterBuilder builder(*model, resolver);
    builder(&interpreter);

    if (!interpreter)
    {
        cerr << "Failed to create interpreter" << endl;
        return -1;
    }
    interpreter->AllocateTensors();

    cv::VideoCapture cap(args.videoPath);
    if (!cap.isOpened())
    {
        cerr << "Error: Could not open video." << endl;
        return -1;
    }

    auto start = chrono::high_resolution_clock::now();
    cv::Mat frame;
    while (cap.read(frame))
    {
        // run inference
        vector<cv::Rect> boxes = runInference(frame, interpreter, args.confidenceThreshold);

        // draw bounding boxes
        for (const auto &rect : boxes)
        {
            cv::rectangle(frame, rect, cv::Scalar(0, 255, 0), 2);
        }

        if (args.delay > 0)
        {
            cv::imshow("YOLOv8 License Plate Detection", frame);

            
            if (cv::waitKey(args.delay) == 'q')
            {
                break;
            }
        }

        frame.release();
    }
    auto end = chrono::high_resolution_clock::now();
    cout << "Inference completed in " << chrono::duration_cast<chrono::milliseconds>(end - start).count() / 1000.0f << " seconds." << endl;

    cap.release();
    cv::destroyAllWindows();

    return 0;
}

File: yolov8/inference_tflite.py
import tensorflow as tf
import cv2
import numpy as np
import torch
import argparse
import time

def boxes(prediction, conf_thres=0.25):
    xc = prediction[:, 4:5].amax(1) > conf_thres  # candidates

    prediction = prediction.transpose(-1, -2)  # sort values in the tensor
    y = torch.empty_like(prediction[..., :4])
    xy = prediction[..., :2]  # centers
    wh = prediction[..., 2:4] / 2  # half width-height
    y[..., :2] = xy - wh  # top left xy
    y[..., 2:] = xy + wh  # bottom right xy
    x = y[0, xc[0]]       # pick based on confidence
    return x

parser = argparse.ArgumentParser()
parser.add_argument("-m", "--model", type=str, default="models/yolov8m_plates_e25.tflite", help="Path to the model file")
parser.add_argument("-v", "--video", type=str, default="datasets/plates/test_img_per_frame.mp4", help="Path to the video file")
parser.add_argument("-d", "--delay", type=int, default=1, help="Delay between frames, 0 means do not show output video.")
parser.add_argument("-c", "--confidence", type=float, default=0.5, help="Confidence threshold")
parser.add_argument("-q", "--quantized", action="store_true", help="Quantized model")

args = parser.parse_args()

quantized = "quantized" in args.model or args.quantized

# load TFLite model and allocate tensors.
interpreter = tf.lite.Interpreter(model_path=args.model)
interpreter.allocate_tensors()

# get input and output tensors.
input_details = interpreter.get_input_details()
output_details = interpreter.get_output_details()
print(input_details)
print()
print(output_details)

cap = cv2.VideoCapture(args.video)

if not cap.isOpened():
    print("Error: Could not open video.")
    exit()

start = time.time()
try:
    while True:
        ret, frame = cap.read()
        if not ret:
            break
        
        start_sample = time.time()

        if quantized:
            frame_normalized = frame.astype(np.int16) - 128
            inputs = np.zeros((1, 3, 640, 640), dtype=np.int8)
            input_data = tf.cast(frame_normalized, tf.int8)
        else:
            # normalize the frame
            frame_normalized = frame / 255.0
            inputs = np.zeros((1, 3, 640, 640), dtype=np.float32)
            input_data = tf.cast(frame_normalized, tf.float32)


        # convert the image from WHC to CHW
        inputs[0, 0] = input_data[:, :, 2]
        inputs[0, 1] = input_data[:, :, 1]
        inputs[0, 2] = input_data[:, :, 0]

        # set the input tensor
        interpreter.set_tensor(input_details[0]['index'], inputs)

        # perform inference
        interpreter.invoke()

        # get the output tensor
        output_data = interpreter.get_tensor(output_details[0]['index'])
        output_data = torch.from_numpy(output_data)
        results = boxes(output_data, args.confidence)
        
        for result in results: # draw bounding boxes
            x1, y1, x2, y2 = map(int, result[:4])
            print(f"Detected box at: ({x1}, {y1}) ({x2}, {y2})")
            cv2.rectangle(frame, (x1, y1), (x2, y2), (0, 255, 0), 2)
        
        end_sample = time.time()
        print(f"Processed frame in {(end_sample - start_sample) * 1000:.3f} ms.", end="\n\n")
        
        if args.delay > 0:
            cv2.imshow("YOLOv8 License Plate Detection", frame)

            if cv2.waitKey(1) & 0xFF == ord('q'):
                break
except KeyboardInterrupt:
    pass
end = time.time()
print(f"\nInference completed in {end - start:.4f} s.")
